src/main/java/server/response/PathFinder.java

```diff
@@ -0,0 +1,41 @@
+package main.java.server.response;
+
+import main.java.application.App;
+import main.java.server.request.Request;
+
+public class PathFinder {
+    private final App app;
+    private Request request;
+
+    public PathFinder(Request request, App app) {
+        this.request = request;
+        this.app = app;
+    }
+
+    public Response selectHandler() {
+
+         if (!pathExists()) {
+             ResponseBuilder builder = new ResponseBuilder();
+             return builder.notFound(request, app.routes());
+         }
+
+        if (!methodAllowed()) {
+            ResponseBuilder builder = new ResponseBuilder();
+            return builder.notAllowed(request, app.routes());
+        }
+
+        return app.routes().get(request.getRequestPath()).get(request.getRequestMethod()).generateResponse(request);
```

Comment:
https://en.wikipedia.org/wiki/Law_of_Demeter 😄

Reply:
Question on this one: Is something like the following an appropriate fix?
`TreeMap<String, HashMap<String, Handler>> routes = app.routes();`
`HashMap<String, Handler> handlerCollection = routes.get(request.getRequestPath());`
`Handler routeHandler = handlerCollection.get(request.getRequestMethod());`
`return routeHandler.generateResponse(request);`

Meaning, just break up the indexing into the data structure into separate steps.

OR

Is the idea that the way I have it now (both the initial line you commented on and the separated version above) indicates that `PathFinder` knows a lot about the internals of `app.routes`, and to fix the Demeter violation I should be encapsulating the data structure inside `app.routes` with getters?

Reply:
Good question! You're definitely on to something with option 2 - it would be easier to change the internal implementation of your dependencies if this namespace wasn't using accessors on the primitive data structures they own. That said, option 1 might also be sufficient - since that'll make it straightforward to implement option 2 if you need to.

One other thing that occurs to me while looking at this - it's perhaps interesting that 3 out of the 4 function calls here are taking data from the request as the argument. Just spitballing - but maybe it would make sense for the `PathFinder` to be an implementation detail of the router? And then your main loop could just call `router.getResponse(request)`?

Reply:
That's an interesting thought, it would definitely make it more concise for the router to hold the `routes` data structure and the function that searches for a matching path.